### PR TITLE
feat: add concurrent upscale (MAPCO-3529)

### PR DESCRIPTION
**Pull request by asafMasa**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✖                                                                       |

Support concurrent get tiles for upscale

### CODE REVIEW

File: MergerLogic/DataTypes/Data.cs [hunk 1]
```diff
@@ -198,25 +199,58 @@ protected virtual Extent GetExtent()
...
+                List<Coord> coordsList = new List<Coord>(MaxZoomRead - (MaxZoomRead - coords.Z));
+                for (int i = z - 1; i >= 0; i--)
+                {
+                    baseTileX >>= 1; // Divide by 2
+                    baseTileY >>= 1; // Divide by 2
 
-            Tile? lastTile = null;
+                    coordsList.Add(new Coord(i, baseTileX, baseTileY));
+                }
+                var response = this.InternalGetExistingTile(coordsList.ToArray());
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Data.cs:219]: Instead of adding this function please move it's code to here and remove it.

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: MergerLogic/DataTypes/Data.cs [hunk 2]
```diff
@@ -199,27 +200,52 @@ protected virtual Extent GetExtent()
...
                 baseTileX >>= 1; // Divide by 2
                 baseTileY >>= 1; // Divide by 2
 
-                lastTile = this.Utils.GetTile(i, baseTileX, baseTileY);
-                if (lastTile != null)
+                coordsList.Add(new Coord(i, baseTileX, baseTileY));
+            }
+
+            // Async method to request all tiles that can be used for "upscale" concurrently
+            var responseOfGetlastTileAsync = async delegate (Coord[] coordsArray)
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Data.cs:219]: Rename to `getUpscaleTiles`.

File: MergerLogic/DataTypes/Data.cs [hunk 3]
```diff
@@ -199,27 +200,52 @@ protected virtual Extent GetExtent()
...
-            string message = lastTile == null ? "null" : $"z:{lastTile.Z}, x:{lastTile.X}, y:{lastTile.Y}";
-            this._logger.LogDebug($"[{MethodBase.GetCurrentMethod().Name}] ended, lastTile: {message}");
-            return lastTile;
+                // Get first valid tile that can be upscaled
+                List<KeyValuePair<int, Tile?>> list = new List<KeyValuePair<int, Tile?>>(zOrderToTileDictionary.ToArray());
+                var orderedList = list.OrderBy(kvp => kvp.Key);
+                Tile? lastTile = orderedList.Last().Value;
+                string message = lastTile == null ? "null" : $"z:{lastTile.Z}, x:{lastTile.X}, y:{lastTile.Y}";
+                this._logger.LogDebug($"[{MethodBase.GetCurrentMethod().Name}] ended, lastTile: {message}");
+                return lastTile;
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Data.cs:245]: Move outside to the function.

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: MergerLogic/DataTypes/Data.cs (hunk 2 above)
Comment (asafMasa) [reply MergerLogic/DataTypes/Data.cs:219]: fixed

File: MergerLogic/DataTypes/Data.cs (hunk 3 above)
Comment (asafMasa) [reply MergerLogic/DataTypes/Data.cs:245]: done

File: MergerLogic/DataTypes/Data.cs (hunk 1 above)
Comment (asafMasa) [reply MergerLogic/DataTypes/Data.cs]: fixed

**Pull request merged**